source/MissionTimer.cpp

```diff
@@ -0,0 +1,386 @@
+		return;
+
+	// Determine if the player meets the activation requirements for this timer.
+	if(!CanActivate(flagship, player))
+	{
+		Deactivate(player, ui, mission);
+		return;
+	}
+
+	// This timer is now active and should advance its counter by 1 tick.
+	// If the full wait time has elapsed, this timer is complete.
+	isActive = true;
+	if(++timeElapsed >= waitTime)
+	{
+		auto it = actions.find(TimerTrigger::TIMEUP);
+		if(it != actions.end())
+			it->second.Do(player, ui, &mission);
+		isComplete = true;
+	}
+}
+
+
+
+bool MissionTimer::CanActivate(const Ship *flagship, const PlayerInfo &player) const
+{
+	// If this timer has no requirements to check, then it should be active.
+	if(!hasRequirements)
+		return true;
+
+	// Does the player's system match the system filter?
+	if((system && flagship->GetSystem() != system) ||
+			(!systems.IsEmpty() && !systems.Matches(flagship->GetSystem())))
+		return false;
+
+	// Does this timer require that the player is solo (i.e. there are
+	// no escorts in the system with the player)?
+	if(requireSolo)
+	{
+		const System *flagshipSystem = flagship->GetSystem();
+		for(const auto &escort : player.Ships())
+		{
+			// Using GetSystem instead of GetActualSystem so that docked
+			// fighters on the player's flagship don't count against them.
+			if(escort.get() != flagship && !escort->IsParked() && !escort->IsDestroyed()
+					&& escort->GetSystem() == flagshipSystem)
+				return false;
+		}
+	}
+
+	// Does this timer require that the player is idle?
+	if(requireIdle)
+	{
+		// The player can't be sending movement commands.
+		if(!flagship->IsThrusting() && !flagship->IsSteering() && !flagship->IsReversing())
+			return false;
+		// And their ship's velocity must be below the max speed threshold.
+		if(flagship->Velocity().LengthSquared() < idleMaxSpeed)
+			return false;
+	}
+
+	// Does this timer require that the player is peaceful?
+	if(requirePeaceful)
+	{
+		// If the player is required to be peaceful, then none of their weapons
+		// can have a fire command. Special weapons like anti-missile turrets which
+		// are only used defensively and automatically do not count against the
+		// player.
+		for(const Hardpoint &hardpoint : flagship->Weapons())
+			if(!hardpoint.IsSpecial() && hardpoint.WasFiring())
+				return false;
+	}
+
+	// Does this timer require that the player is cloaked or uncloaked?
+	double cloaking = flagship->Cloaking();
+	if((requireUncloaked && cloaking) || (requireCloaked && cloaking != 1.))
+		return false;
+
+	return true;
+}
+
+
+
+void MissionTimer::Deactivate(PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	// If the timer wasn't active the frame before, don't do anything.
+	if(!isActive)
+		return;
+	isActive = false;
+
+	// Reset the timer if it isn't set to pause when deactivated.
+	if(!pauses)
+		timeElapsed = 0;
+
+	// Perform the DEACTIVATION action, if there is one, assuming it hasn't fired yet.
+	if(!triggeredActions.contains(TimerTrigger::DEACTIVATION))
```

Comment:
Let's not search the set twice.
```suggestion
	if(triggeredActions.insert(TimerTrigger::DEACTIVATION).second)
```